featuretools/feature_discovery/feature_discovery.py

```diff
@@ -0,0 +1,413 @@
+    else:
+        keys.append(ANY)
+
+    return keys
+
+
+def index_input_set(input_set: List[ColumnSchema]) -> Dict[str, int]:
+    """
+    Indexes input set to find types of columns and the quantity of eatch
+
+    Args:
+        input_set (List(ColumnSchema)):
+            List of Column types needed by associated primitive.
+
+    Returns:
+        Dict[str, int]
+            A hashmap from key to int
+
+    Examples:
+        .. code-block:: python
+
+            from featuretools.feature_discovery.feature_discovery import get_features
+            from woodwork.column_schema import ColumnSchema
+
+            input_set = [ColumnSchema(semantic_tags={"numeric"}), ColumnSchema(semantic_tags={"numeric"})]
+            indexed_input_set = index_input_set(input_set)
+            {"numeric": 2}
+    """
+    out = {}
+    for column_schema in input_set:
+        for key in column_to_keys(column_schema):
+            out[key] = out.get(key, 0) + 1
+    return out
+
+
+def get_features(
+    feature_groups: Dict[str, List[Feature]],
+    input_set: List[ColumnSchema],
+    commutative: bool,
+) -> List[List[Feature]]:
+    """
+    Calculates all Feature combinations using the given hashmap of existing features, and the input set of required columns.
+
+    Args:
+        feature_groups (Dict[str, List[Feature]]):
+            Hashmap from Key to List of Features. Key is either: LogicalType name (eg. "Double"), Semantic tag (eg. "index"),
+            or combination (eg. "Double,index").
+        input_set (List(ColumnSchema)):
+            List of Column types needed by associated primitive.
+        commutative (bool):
+            whether or not we need to use product or combinations to create feature sets.
+
+    Returns:
+        List[List[Feature]]
+            A list of Feature sets.
+
+    Examples:
+        .. code-block:: python
+
+            from featuretools.feature_discovery.feature_discovery import get_features
+            from woodwork.column_schema import ColumnSchema
+
+            feature_groups = {
+                "ANY": ["f1", "f2", "f3"],
+                "Double": ["f1", "f2", "f3"],
+                "numeric": ["f1", "f2", "f3"],
+                "Double,numeric": ["f1", "f2", "f3"],
+            }
+            input_set = [ColumnSchema(semantic_tags={"numeric"}), ColumnSchema(semantic_tags={"numeric"})]
+            features = get_features(col_groups, input_set, commutative=False)
+    """
+    indexed_input_set = index_input_set(input_set)
+
+    prod_iter = []
+    for name, count in indexed_input_set.items():
+        if name not in feature_groups:
+            return []
+
+        features2 = feature_groups[name]
+
+        if commutative:
+            prod_iter.append(combinations(features2, count))
+        else:
+            prod_iter.append(permutations(features2, count))
+
+    feature_combinations = product(*prod_iter)
+
+    return [[z for y in x for z in y] for x in feature_combinations]
```

Comment:
This is a little confusing to me -- can we make the variable names more descriptive?